### ISSUE TITLE
Modifying PeLoader and format.pe.* to support loading images into existing programs

### ISSUE THREAD

**Issue by Rantanen** [NationalSecurityAgency/ghidra#2571]
TLDR; Is Ghidra interested in couple of PRs to open up `PeLoader`/`PortableExecutable` APIs to support loading (multiple) PE images to arbitrary base addresses within a `Program`?

---

I've been working in [Windows Minidump loader](https://github.com/Rantanen/ghidra-minidump-loader) for Ghidra. The Minidumps contain the full PE images for loaded modules (with slight differences in memory layout) and as discussed in #2552, the easiest way to get most out of these was to delegate the actual image loading to the `PeLoader`.

Things seem to be working now well enough that I'm shifting off the proor-of-concept phase and starting to think about releases and maintenance of the project. There were couple of changes I had to make to `PeLoader`, mostly to support loading PE images into an existing program and into a different address than `Program.baseImage`. Currently I'm maintaining a duplicate of the `PeLoader` and `AbstractPeDebugLoader` classes as well as the whole `ghidra.app.util.bin.format.pe` package in tree in my project.

Would Ghidra be interested in pull requests to incorporate similar changes into Ghidra itself? This would allow plugins that need to load PE images into existing programs benefit from `PeLoader` - although I'm not sure if this applies to anything but the Minidump loader.

I don't think the individual changes themselves are too bad if introduced one by one. If Ghidra is interested in the changes, I'll want to discuss them separately in PRs to ensure they are properly done and match Ghidra's architecture, but for demonstration purposes, the current changes (that I can remember) are:

- Introduce a new constructor for `PeLoader` that can be used to parametrize the instance to use `MEMORY` layout instead of `FILE` layout.

- Expose the `load` method publicly.

- Support the executable-specific base address when processing the headers.
  - There are couple of options for how this could be done. My current implementation wraps the `SectionLayout` and new `imageBaseAddress` into `ImageLoadInfo` type that is passed in by the loader. This also carries things like `moduleName`, as it now differs from the `Program.getName`.
  - An alternative that would work for me is to use the `SectionLayout` to figure out whether the addresses should be calculated off `Program.imageBase()` or `OptionalHeader.imageBase()`; As Windows loads the executable into memory it sets the image base into the real image location - although if there's a need to load `SectionLayout.MEMORY` images into arbitrary base addresses this approach would prevent that.

- Split the loading into two methods to support mapping all images into memory before processing them:
  - `loadImage` that just loads the image into memory in the correct position.
  - `processImage` that performs the header processing, marks symbols, etc.
  - There will be need to pass some data from `loadImage` to `processImage`, my current approach is to use a simple struct as a return type that combines the necessary variables.

- Link imported functions directly if they are loaded.
  - Minidumps contain the runtime memory so the imported functions have been already resolved and are pointing to the bound locations.
  - When marking imported functions, the imports check whether the target address is present in `Program.getMemory()` and if so, it creates a direct memory reference as opposed to an external reference.
  -  (This is the reason why I needed to load all images into memory before processing them.)

- Introduce `ModuleBaseOffset32`/`ModuleBaseOffset64` types to be used when the PE image is not at `Program.imageBase`.
  - The PE headers contain couple of relative virtual addresses (RVAs) that are based on the PE image base. `ImageBaseOffset..` types work for images that are loaded at `Program.imageBase`, but loading the PE image somewhere else breaks these data types.
  - `ModuleBaseOffset..` are equivalent to `ImageBaseOffset..` types with the exception that the base address is now the base address of the containing module.
  - What is "the base address of the containing module"? Good question! Currently the Minidump loader marks this in the `ProgramUserData` and the `ModuleBase..` offsets look up the "closest previous" module boundary user data. Unfortunately this is the best implementation I could come up with while still sticking to data types instead of just making explicit memory references.
  - Should probably default to `imageBase` if module base is not found.

These changes are in somewhat separate commits in the [Windows Minidump loader](https://github.com/Rantanen/ghidra-minidump-loader/commits/main) repository history if someone wants to have a look at them. Though as a warning, they were made with no input from anyone who actually knows the code base and they have not necessarily been fully cleaned up:

As an example the initial "have everything based on the PE image offset" change had me add that offset as a parameter to `PeMarkupable`, which caused a ton of (small) changes. Later I realized that `NtHeader` needs it anyway and that type is already carried in the calls through `PeMarkupable` so I could just expose it in `NtHeader` and get rid of the parameter I added. Having it in one place (`NtHeader`) also prevents it getting out of sync and having parts of the program loaded with different load info.

Fear not, if Ghidra is interested in changes like these, any possible PRs would be written somewhat from scratch to ensure they abide by Ghidra's style guidelines, etc. :)

(I'm trying to keep a door open from my side for the possibility of contributing the Minidump loader directly to Ghidra if there's interest for it once it matures and gets cleaned up. I feel having a source tree that isn't 80% duplicate code from Ghidra would be helpful if that ever happens!)

**Comment by ghidra1** [comment]
Due to the complications of relocation processing and coping with memory conflicts it was decided long ago to not support add-to-program for such binary formats.  You can however accomplish this in a more surgical fashion by importing to a separate program, export to XML, then perform an Add To Program... selecting the XML.  Options can be specified to restrict imported elements.

**Comment by Rantanen** [comment]
I'm not trying to "Add to Program" different PE executables or dlls together. I guess I could have chosen the title a bit better. :)

The Windows Minidump (being somewhat equivalent to a core dump on Linux) contains a snapshot of the process memory, which already includes the PE-images as loaded into memory. The `ghidra-minidump-loader` project already works for loading a Minidump and creating a Ghidra `Program` that contains all PE-images in one memory space with exports/etc. addresses resolved (since the OS has done that when it has loaded the PE images).

The implementation relies heavily on slightly modified Ghidra [`PELoader`, `PortableExecutable` and associated classes](https://github.com/Rantanen/ghidra-minidump-loader/blob/c1d17ee5b3317f4b3694839d28529e36e386bc3a/MinidumpLoader/src/main/java/net/jubjubnest/minidump/contrib/pe/ArchitectureDataDirectory.java) in parsing the PE and loading it in memory (figuring out sections, identifying Import/Export/Debug/etc. directories within the PE image, etc.). The modifications revolve mostly around opening up `PELoader` private methods and the ability to specify "custom" `imageBase` instead of assuming it is always at `Program.imageBase()`.

The benefit of loading a Windows Minidump directly into Ghidra (instead of carving the executables/libraries out of it and loading those individually) is the ability to load everything into the same memory space and the inclusion of the runtime program memory (thread stacks + heap).

The underlying question is whether Ghidra is interested in upstreaming the whole Minidump Loader in the future when it has matured enough.

If Ghidra is interested in having the Minidump Loader as part of its built-in loaders in the future, I'd like to avoid radical changes in my forks of the `PELoader`/`PortableExecutable` to avoid divergence (and thus make the possible future PR easier); If there's no interest, then it's easier for me to be a bit more heavy handed with my changes to the forked types.

The reason I approached this from the point of `PELoader` and `format.pe.*` (ie. `PortableExecutable`) changes was that submitting smaller PRs would allow me to get more familiar with Ghidra's style preferences, which would have made it easier to keep the actual Minidump Loader codebase closer to those same preferences as well. But this is also relevant only in case I'll end up submitting a PR for the Minidump Loader for Ghidra - which depends entirely on whether Ghidra is interested in said Loader in the first place. :)

I do acknowledge that I'm using Ghidra more or less as a debugger and this use case might not be that relevant for common reverse engineering scenarios, which is why it's hard for me to judge whether the use case is of interest to Ghidra.

(This might also apply to PDB-related functionality, which I'm currently working on and seems to involve re-implementing a good part of `PdbSymbolServerPlugin` functionality.)

**Comment by ghidra1** [comment]
Sorry for the confusion.  I think I got my wires crossed.  I removed my previous comment because it was way off topic.

**Comment by ghidra1** [comment]
While I have not digested everything above, I was unclear why you chose to use `ProgramUserData` for storage which is independent of the Program database storage (e.g., is not shared when multi-user repository used).

**Comment by Rantanen** [comment]
> Sorry for the confusion. I think I got my wires crossed. I removed my previous comment because it was way off topic.

It made somewhat sense for the previous title, which I've since changed so the confusion was partly on me as well, sorry. :)

> While I have not digested everything above, I was unclear why you chose to use `ProgramUserData` for storage which is independent of the Program database storage (e.g., is not shared when multi-user repository used).

I've been a bit wary of opening a pile of Question-issues here to avoid spamming all the maintainers (or any who have notifications enabled anyway!) so many of my current decisions are based on "this is the first thing that seemed to do what I want it to do"! `ProgramUserData` was the first concept that I found, which allowed me to "pass information" from the Loader to the tools/plugins/etc. Especially information that was tied to addresses and could be iterated over. Didn't know it wasn't shared in a multi-user repository! Sounds like I'll need to find some other approach for it.

Thanks for that notice!

(Just noted Ghidra has enabled the GitHub discussions, which sounds like the perfect place for clarifying few of my concerns!)

**Comment by ghidra1** [comment]
See discussion on program related data storage capabilities: #2593

**Comment by Rantanen** [comment]
Thanks! That's extremely helpful! I was still typing out my own question. :)

**Comment by Rantanen** [comment]
Sounds like the `getUsrPropertyManager` is the way to go for me! The naming of that function made me believe it was also specific to each user, but the note about possible conflicts makes me now believe that's not the case and instead it's common to all users.

**Comment by ghidra1** [comment]
I think User was meant to indicate that they are for adhoc use by user scripts and plugins.  But the naming on many classes and methods leave a lot of room for improvement :)

**Comment by dev747368** [comment]
> (This might also apply to PDB-related functionality, which I'm currently working on and seems to involve re-implementing a good part of `PdbSymbolServerPlugin` functionality.)

I would hold off on that if I were you.  We're working on a revamp for Pdb symbol server searching and any changes you make probably would conflict.

**Comment by Rantanen** [comment]
Too late! :)

Pending resolution to the question of whether I should attempt to re-use Ghidra's internals in my implementation or not, I've ended up re-implementing most of the PDB resolution (symbol server searching mainly) instead of adapting the internals (like I did with PeLoader/PortableExecutable). If the revamped symbol server implementation fits my needs, I'll just scrap my current implementation and move on top of the new one.

Most of Ghidra's PDB-functionality was almost reusable for me, but there were some slight issues, such as `PdbLocator` accepting `PdbProgramAttributes` as a parameter for `findPdb` but then almost immediately resolving it again through `Program` which will fail in my use case. The change to pass the received `PdbProgramAttributes` through one function call to avoid having to resolve it again would have been minor, but given the uncertainty with this issue, I avoided forking any more Ghidra internals and just re-implemented those parts of PDB resolution myself.

The one piece I don't want to re-implement is the `PdbApplicator`, but fortunately the `applyTo` method already accepts `Address imageBaseAddress` as a parameter, which allows me to reuse it once the PDB file is identified and parsed. <3

**Comment by dev747368** [comment]
@Rantanen fyi, our rework of the PDB symbol server stuff landed in commit 22675e9f39dd23bb1f0df775ef3cc2c8f22e81a7 if you want to take at look at it.
